crates/typst/src/geom/gradient.rs

```diff
@@ -0,0 +1,1089 @@
+        }
+        .clamp(0.0, 1.0)
+    }
+}
+
+/// Pre-processes the stops, checking that they are valid and computing the
+/// offsets if necessary.
+///
+/// Returns an error if the stops are invalid.
+///
+/// This is split into its own function because it is used by all of the
+/// different gradient types.
+fn process_stops(stops: &[Spanned<Stop>]) -> SourceResult<Vec<(Color, Ratio)>> {
+    let has_offset = stops.iter().any(|stop| stop.v.offset.is_some());
+    if has_offset {
+        let mut last_stop = NEG_INFINITY;
+        for Spanned { v: stop, span } in stops.iter() {
+            let Some(stop) = stop.offset else {
+                bail!(error!(
+                    *span,
+                    "either all stops must have an offset or none of them can"
+                )
+                .with_hint("try adding an offset to all stops"));
+            };
+
+            if stop.get() < last_stop {
+                bail!(*span, "offsets must be in stricly monotonic");
+            }
+
+            last_stop = stop.get();
+        }
+
+        return stops
+            .iter()
+            .map(|Spanned { v: Stop { color, offset }, span }| {
+                if offset.unwrap().get() > 1.0 || offset.unwrap().get() < 0.0 {
+                    bail!(*span, "offset must be between 0 and 1");
+                }
+                Ok((*color, offset.unwrap()))
+            })
+            .collect::<SourceResult<Vec<_>>>();
+    }
+
+    Ok(stops
+        .iter()
+        .enumerate()
+        .map(|(i, stop)| {
+            let offset = i as f64 / (stops.len() - 1) as f64;
+            (stop.v.color, Ratio::new(offset))
+        })
+        .collect())
+}
+
+/// Sample the stops at a given position.
+fn sample_stops(stops: &[(Color, Ratio)], mixing_space: ColorSpace, t: f64) -> Color {
+    let t = t.clamp(0.0, 1.0);
+    let mut low = 0;
+    let mut high = stops.len();
+
+    while low < high {
+        let mid = (low + high) / 2;
+        if stops[mid].1.get() < t {
+            low = mid + 1;
+        } else {
+            high = mid;
+        }
+    }
+
+    if low == 0 {
+        low = 1;
+    }
+    let (col_0, pos_0) = stops[low - 1];
+    let (col_1, pos_1) = stops[low];
+    let t = (t - pos_0.get()) / (pos_1.get() - pos_0.get());
+
+    Color::mix(
+        vec![WeightedColor::new(col_0, 1.0 - t), WeightedColor::new(col_1, t)],
+        mixing_space,
+    )
+    .unwrap()
+}
+
+fn cubehelix_to_rgb(h: f32, s: f32, l: f32) -> Color {
+    let h = (h + 120.0).to_radians();
+    let l = l;
+    let a = s * l * (1.0 - l);
+
+    let (sinh, cosh) = h.sin_cos();
+
+    let r = l - a * (0.14861 * cosh - 1.78277 * sinh).min(1.0);
+    let g = l - a * (0.29227 * cosh + 0.90649 * sinh).min(1.0);
+    let b = l + a * (1.97294 * cosh);
+
+    Color::Rgba(Rgba::new(r.clamp(0.0, 1.0), g.clamp(0.0, 1.0), b.clamp(0.0, 1.0), 1.0))
+}
+
+macro_rules! preset {
+    ($name:ident; $($colors:literal),* $(,)*) => {
+        #[comemo::memoize]
```

Comment:
A once cell `Lazy` would probably be faster. Probably not particular performance-sensitive though.

Reply:
I did the change anyway :)